Parameter does not escape annotation (#3961)
dotnet/csharplang: proposals/low-level-struct-improvements.md

        {
            ...
        })
    }
}
```

In order to work around this low level code will resort to `unsafe` tricks to 
lie to the compiler about the lifetime of their `ref struct`. This significantly
reduces the value proposition of `ref struct` as they are meant to be a means 
to avoid `unsafe` while continuing to write high performance code.

The other place where parameter default escape scope causes friction is when 
they are re-assigned within a method body. For instance if a method body decides
to conditionally apply escaping to input by using stack allocated values. Once
again this runs into some friction.

```cs
void WriteData(ReadOnlySpan<char> data)
{
    if (data.Contains(':'))
    {
        Span<char> buffer = stackalloc char[256];
        Escape(data, buffer, out var length);

        // Error: Cannot assign `buffer` to `data` here as the *safe-to-escape*
        // scope of `buffer` is to the current method scope while `buffer` is
        // outside the current method scope
        data = buffer.Slice(0, length);
    }

    WriteDataCore(data);
}
```

This pattern is fairly common across .NET code and it works just fine when 

> `ref struct` are not involved. Once users adopt `ref struct` though it forces them

Reviewer: ```suggestion
a `ref struct` is not involved. Once users adopt `ref struct` though it forces them
```

to change their patterns here and often they just resort to `unsafe` to work
around the limitations here.

To remove this friction the language will provide the attribute 
`[DoesNotEscape]`. This can be applied to parameters of any type or instance 
members defined on `ref struct`. When applied to parameters the *safe-to-escape*
and *ref-safe-to-escape* scope will be the current method scope. When applied to
instance members of `ref struct` the same limitation will apply to the `this`
parameter.

```cs
class C
{

>     static Span<int> M1(Span<int> p1, [DoesNotEsacpe] Span<int> p2)

Reviewer 2: ```suggestion
    static Span<int> M1(Span<int> p1, [DoesNotEscape] Span<int> p2)
```

`ref struct` that are assigned on multiple paths where at least one of the 
paths is a limited *safe-to-escape* scope. 

```cs
int length = ...;
Span<byte> span;
if (length > StackAllocLimit)
{
    span = new Span(new byte[length]);
}
else
{
    // Error: The *safe-to-escape* of `span` was decided to be outside the 
    // current method scope hence it can't be the target of a stackalloc
    span = stackalloc byte[length];
}
```

For `Span<T>` specifically developers can work around this by initializing the 
local with a `stackalloc` of size zero. This changes the *safe-to-escape* scope
to be the current method and is optimized away by the compiler. It's effectively
a syntax for making a `[DoesNotEscape]` local.

```cs
int length = ...;
Span<byte> span = stackalloc byte[0];
if (length > StackAllocLimit)
{
    span = new Span(new byte[length]);
}
else
{
    // Okay
    span = stackalloc byte[length];
}
```

This only works for `Span<T>` though, there is no general purpose mechanism for
`ref struct` values. However the `[DoesNotEscape]` attribute provides exactly 
the semantics that are desired here. If we decide in the future to allow 

> attributes to apply to local functions it would provide immediate relief to this

Reviewer: We allow attributes to apply to local functions.

Author: Blah, I meant to say local variables .

> attributes to apply to local functions it would provide immediate relief to this

Author: ```suggestion
attributes to apply to local variables it would provide immediate relief to this
```
